### PR TITLE
Reduce overhead of broadcasting when broadcasting isn't required.

### PR DESCRIPTION
**Pull request by gchanan**
Essentially if the arguments are all the same size, we skip the expand and use the original tensor.

### CODE REVIEW

File: tools/cwrap/plugins/Broadcast.py [hunk 1]
```diff
@@ -75,29 +90,44 @@ def getExpandTemplate(self, expand_call, success_code, raise_errors):
...
-
-    def getOutPlacePreExpand2Template(self, raise_errors):
-        expand_code = """expand_outplace2(LIBRARY_STATE ${arg_op_a}_guard.get(), ${arg_op_other}_guard.get(),
-                                          ${arg_op_a}, ${arg_op_other},
-                                          \"${op_a}\", \"${op_other}\", !${raise_errors});"""
+                success_code + "\n"
+                "}\n")
+
+    def getOutPlacePreExpand2Template(self, type_op_a, type_op_other, raise_errors):
+        size_check = """THSize_isSameSizeAs(${arg_op_a}->size, ${arg_op_a}->nDimension,
```
Comment (ngimel) [inline tools/cwrap/plugins/Broadcast.py]: Should you also be comparing strides, not just size?

**Comment by gchanan** [comment]
Addresses at least some of https://github.com/pytorch/pytorch/issues/2012.

Here's a microbenchmark (modeled after @apaszke's) that measures 4 common cases (outplace vs inplace, 2 args vs 3 args):
```
import torch
from timeit import timeit

x = torch.randn(5, 5)
y = torch.randn(5, 5)
z = torch.randn(5, 5)

print("outplace2", timeit('x + y', number=1000000, globals=locals()))
print("outplace3", timeit('x.addcmul(y,z)', number=1000000, globals=locals()))
print("inplace1", timeit('x.add_(y)', number=1000000, globals=locals()))
print("inplace2", timeit('x.addcmul_(y,z)', number=1000000, globals=locals()))
```

without broadcasting code:
```
outplace2 1.170712536200881
outplace3 0.6207181625068188
inplace1 0.28007462061941624
inplace2 0.2738169729709625
```

with existing broadcasting code:
```
outplace2 2.0662140138447285
outplace3 1.7219338584691286
inplace1 0.703955814242363
inplace2 0.9586600493639708
```

with new broadcasting code:
```
outplace2 1.1887976974248886
outplace3 0.6337768454104662
inplace1 0.27564410865306854
inplace2 0.29433456622064114
```

Note that this is potentially slower in the case where broadcasting is required, since we do an extra shape check before executing the old code.  Obviously, this could optimized with some extra work, but it's already probably a good tradeoff given broadcasting is probably relatively rare.  To compare the case where broadcasting is enabled, change the benchmark line that sets y to:
```
y = torch.randn(5)
```

Here are the results then:
before change:
```
outplace2 1.8439689576625824
outplace3 1.8589385487139225
inplace1 0.7429333087056875
inplace2 1.070244874805212
```

after change:
```
outplace2 1.841021191328764
outplace3 1.8419923204928637
inplace1 0.7541772909462452
inplace2 1.0395784191787243
```

so it doesn't seem to make much of a difference and it's probably not worth further optimizing before getting this in.

File: tools/cwrap/plugins/Broadcast.py (hunk 1 above)
Comment (gchanan) [inline tools/cwrap/plugins/Broadcast.py]: It's not necessary; this is just checking if we can early-opt out of the broadcasting code, so it works as if the broadcasting code wasn't there.  So, assuming these operations worked on tensors of same size but different strides, this code should work as well.

**Comment by killeent** [comment]
@gchanan can you post a snippet of what the generated code looks like?

**Comment by gchanan** [comment]
@killeent There's a few cases, I'll just pick a representative one of an out-of-place 2 tensor operation with matching types (i.e. x + y) and pointwise fallback (i.e. 1-d view semantics).  Also I'm only showing the pre function call code for simplicity.

Previously:
```
THTensorPtr arg_self_guard(THTensor_(new)(LIBRARY_STATE_NOARGS));
THTensorPtr arg_other_guard(THTensor_(new)(LIBRARY_STATE_NOARGS));

bool expand_success = false;
try {
   expand_outplace2(LIBRARY_STATE arg_self_guard.get(), arg_other_guard.get(),
         arg_self, arg_other,...)
   expand_success = true;
}
catch (std::exception &e) {}
if(expand_success) {
  arg_self = arg_self_guard.get();
  arg_other = arg_other_guard.get();
}
// call THTensor_(add) as usual...
```

Now:
```
THTensorPtr arg_self_guard(nullptr);
THTensorPtr arg_other_guard(nullptr);

bool try_expand = !THSize_isSameSizeAs(arg_self->size, arg_self->nDimension,
     arg_other->size, arg_other->nDimension);
if (try_expand) {
  bool expand_success = false;
  try {
    arg_self_guard = THTensor_(new)(LIBRARY_STATE_NOARGS);
    arg_other_guard = THTensor_(new)(LIBRARY_STATE_NOARGS);

     expand_outplace2(LIBRARY_STATE arg_self_guard.get(), arg_other_guard.get(),
         arg_self, arg_other,...)
     expand_success = true;
  }
  catch (std::exception &e) {}
   if(expand_success) {
     arg_self = arg_self_guard.get();
     arg_other = arg_other_guard.get();
   }
}
// call THTensor_(add) as usual...
```

So previously we were always creating new tensors and expanding them to the correct size, now we are only doing that if the sizes don't match (there are further optimizations one could make e.g. if self's expanded size is the same as the other's size, we could not expand other's).

**Review by killeent — APPROVED** [review]

**Pull request merged**